core/interval/int.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package interval
+
+import (
+	"sort"
+)
+
+// IntegerInterval represents the integers contained in the mathematical interval
+// [Lower, Higher]. That is, the set of all integers x such that Lower <= x
+// and x <= Upper.
+type IntegerInterval struct {
+	// Lower is the lower bound of the interval (inclusively).
+	Lower int
+
+	// Upper is the upper bound of the interval (inclusively).
+	Upper int
+}
+
+// NewIntegerInterval returns a new IntegerInterval within the given bounds.
+func NewIntegerInterval(a, b int) IntegerInterval {
+	return IntegerInterval{
+		Lower: min(a, b),
+		Upper: max(a, b),
+	}
+}
+
+// Intersects returns true if the two intervals contain any common elements.
+func (i IntegerInterval) Intersects(other IntegerInterval) bool {
+	return i.Lower <= other.Upper && i.Upper >= other.Lower
+}
+
+// Adjacent returns true if the two intervals are adjacent to each other. That
+// is, if the upper bound of one interval is one less than the lower bound of
+// the other.
+//
+// If two intervals are adjacent, they can be merged into a single interval.
+func (i IntegerInterval) Adjacent(other IntegerInterval) bool {
+	return i.Upper+1 == other.Lower || i.Lower-1 == other.Upper
+}
```

Comment:
If we make it specific to ports, we don't have to cater to the generalised stance.